notes/sparse.md

```diff
@@ -7,30 +7,58 @@ sort: 10
 
 * * *
 
-## Dense Matrices
+## Definition
 
-A \\(n \times n\\) matrix is called dense if it has <span>\\(O(n^2)\\)</span> non-zero entries. For example:
-<div>\[\mathbf{A} = \begin{bmatrix} 1.0 & 2.0 & 3.0 \\ 4.0 & 5.0 & 6.0 \\ 7.0 & 8.0 & 9.0 \end{bmatrix}.\]</div>
+A Sparse matrix is a matrix with few non-zero entries.
 
-To store the matrix, all components are saved in row-major order.  For <span>\\(\mathbf{A}\\)</span> given above, we would store:
-<div>\[AA = \begin{bmatrix} 1.0 & 2.0 & 3.0 & 4.0 & 5.0 & 6.0 & 7.0 & 8.0 & 9.0 \end{bmatrix}.\]</div>
+A \\(m \times n\\) matrix is called sparse if it has <span>\\(O(min(m, n))\\)</span> non-zero entries. 
+
+## Goals
+
+A sparse matrix aims to store large matrices efficiently without storing many zeros, which allows for more economical computations.
+
+The number of operations required to add two dense matrices <span>\\(\mathbf{P}\\)</span>, <span>\\(\mathbf{Q}\\)</span> is  <span>\\(O(n(\mathbf{P}) \times nnz(\mathbf{ }))\\)</span> where <span>\\n(\mathbf{X})\\)</span> is the number of elements in <span>\\(\mathbf{X}\\)</span>.
+
+The number of operations required to add two sparse matrices <span>\\(\mathbf{P}\\)</span>, <span>\\(\mathbf{Q}\\)</span> is  <span>\\(O(nnz(\mathbf{P}) \times nnz(\mathbf{Q}))\\)</span> where <span>\\nnz(\mathbf{X})\\)</span> is the number of non-zero elements in <span>\\(\mathbf{X}\\)</span>.
+
+## Storage Solutions
 
-The dimensions of the matrix are stored separately.
+Let's explore ways to store the following example:
+<div>\[\mathbf{A}= \begin{bmatrix} 1.0 & 0 & 0 & 2.0 & 0 \\ 3.0 & 4.0 & 0 & 5.0 & 0 \\ 6.0 & 0 & 7.0 & 8.0 & 9.0 \\ 0 & 0 & 10.0 & 11.0 & 0 \\ 0 & 0 & 0 & 0 & 12.0 \end{bmatrix}.\]</div>
 
-## Sparse Matrices
 
-A \\(n \times n\\) matrix is called sparse if it has <span>\\(O(n)\\)</span> non-zero entries. For example:
+### Dense Matrices
+
+A \\(m \times n\\) matrix is called dense if it stores all values including zeros. To store <span>\\(\mathbf{A}\\)</span> above:
 <div>\[A = \begin{bmatrix} 1.0 & 0 & 0 & 2.0 & 0 \\ 3.0 & 4.0 & 0 & 5.0 & 0 \\ 6.0 & 0 & 7.0 & 8.0 & 9.0 \\ 0 & 0 & 10.0 & 11.0 & 0 \\ 0 & 0 & 0 & 0 & 12.0 \end{bmatrix}.\]</div>
 
-**COO** (Coordinate Format) stores arrays of row indices, column indices and the corresponding non-zero data values in any order. This format provides fast methods to construct sparse matrices and convert to different sparse formats. For <span>\\({\bf A}\\)</span> the COO format is:
+To store the matrix, all components are saved in row-major order.  For <span>\\(\mathbf{A}\\)</span> given above, we would store:
+<div>\[A_{dense} = \begin{bmatrix} 1.0 & 2.0 & 3.0 & 4.0 & 5.0 & 6.0 & 7.0 & 8.0 & 9.0 \end{bmatrix}.\]</div>
+
+The dimensions of the matrix are stored separately:
+<div>\[A_{shape} = \\((nrow, ncol)\\)\]</div>
+
+The benefits of this method include simplicity and easy-blocked format.
```

Comment:
also memory/cache advantages and better vectorization (although numpy doesn't require that rows are block-aligned)